presto-main/src/test/java/com/facebook/presto/sql/planner/TestDereferencePushDown.java

```diff
@@ -0,0 +1,197 @@
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.equiJoinClause;
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.exchange;
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.expression;
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.filter;
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.join;
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.output;
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.project;
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.semiJoin;
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.sort;
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.unnest;
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.values;
+import static com.facebook.presto.sql.planner.plan.ExchangeNode.Scope.LOCAL;
+import static com.facebook.presto.sql.planner.plan.ExchangeNode.Type.GATHER;
+import static com.facebook.presto.sql.planner.plan.ExchangeNode.Type.REPARTITION;
+import static com.facebook.presto.sql.planner.plan.JoinNode.Type.INNER;
+import static com.facebook.presto.sql.tree.SortItem.NullOrdering.LAST;
+import static com.facebook.presto.sql.tree.SortItem.Ordering.ASCENDING;
+
+public class TestDereferencePushDown
+        extends BasePlanTest
+{
+    @Test
+    public void testDereferencePushdownJoin()
+    {
+        assertPlan("WITH t(msg) AS (SELECT * FROM (VALUES ROW(CAST(ROW(1, 2.0) AS ROW(x BIGINT, y DOUBLE))))) " +
+                        "SELECT b.msg.x FROM t a, t b WHERE a.msg.y = b.msg.y",
+                output(ImmutableList.of("b_x"),
+                        join(INNER, ImmutableList.of(equiJoinClause("a_y", "b_y")),
+                                anyTree(
+                                        project(ImmutableMap.of("a_y", expression("msg.y")),
+                                                values("msg"))
+                                ), anyTree(
+                                        project(ImmutableMap.of("b_y", expression("msg.y"), "b_x", expression("msg.x")),
+                                                values("msg"))))));
+
+        assertPlan("WITH t(msg) AS ( SELECT * FROM (VALUES ROW(CAST(ROW(1, 2.0) AS ROW(x BIGINT, y DOUBLE))))) " +
+                        "SELECT a.msg.y FROM t a JOIN t b ON a.msg.y = b.msg.y WHERE a.msg.x > bigint '5'",
+                output(ImmutableList.of("a_y"),
+                        join(INNER, ImmutableList.of(equiJoinClause("a_y", "b_y")),
+                                anyTree(
+                                        project(ImmutableMap.of("a_y", expression("msg.y")),
+                                                filter("msg.x > bigint '5'",
+                                                        values("msg")))
```

Comment:
Shouldn't we have a `Project(msg.x as a_x, msg.y as a_y) ` in between the `filter` and `values` nodes?

Reply:
![image](https://user-images.githubusercontent.com/1274155/84719518-6e2dae80-af30-11ea-9ad8-a774d43590b6.png)

only a_y appears in top ProjectNode, so a_y is pushed down. a_x only appears in FilterNode, it is not pushed down.